servers/rendering/renderer_rd/renderer_scene_render_rd.cpp

```diff
@@ -340,14 +340,16 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 
 	// Glow, auto exposure and DoF (if enabled).
 
-	Size2i internal_size = rb->get_internal_size();
 	Size2i target_size = rb->get_target_size();
-
 	bool can_use_effects = target_size.x >= 8 && target_size.y >= 8; // FIXME I think this should check internal size, we do all our post processing at this size...
 	bool can_use_storage = _render_buffers_can_be_storage();
 
+	bool use_fsr = fsr && can_use_effects && rb->get_scaling_3d_mode() == RS::VIEWPORT_SCALING_3D_MODE_FSR;
```

Comment:
I don't think we can do this any better until we have a proper render graph so we don't need to hardcode this stuff.

But seeing FSR is only embedded in the Forward+ renderer, once we start extending the generic code, this will start getting in the way pretty quickly. We've made for too many assumptions in the rendering code that are now no longer true and we really risk painting ourselves into a corner with more and more exceptions being embedded here as we introduce more effects into the rendering pipeline.

One alternative that may be nicer here but maybe a bit much as a change for FSR, is that we let go of the idea of internal texture as this is way too restrictive.

Instead at the start of rendering we determine our 3d render size (as a replacement of internal size) and we leave it up to `RendererSceneRenderRD::_render_scene` to do what it must do in order to render the 3D output, and it then simply returns the RID and size of whatever end result it has made, whether that is the original 3D render buffer (internal_texture) or the upscaled texture (FSR) or if an effect is run that copies the render output into a new texture, it no longer needs to do a copy back, it is that texture that becomes the output.

It also means that `RenderSceneBuffers` will no longer hardcode some of the buffers we use but again leave it up to the implementation of the render method to request/define required buffers, and in that we'll probably need to provide implementations like sky with the RID of the color texture it needs to render into, instead of it assuming that is the internal texture.

That would add way more flexibility into how we come to our rendering end result, and won't require hardcoding which buffers are being used and make it much easier to chain effects.

Reply:
I agree with the general idea, although that kind of refactor may be out of my scope for now. Should we move it to a PR after this one?

Reply:
Agreed